### PR TITLE
APPMAN-1027 add reservation prior to adding apprenticeship and handler auto deletion

### PR DESCRIPTION
**Pull request by Pauljgraham**

### CODE REVIEW

File: src/Approvals/SFA.DAS.Approvals/Services/AutoReservationsService.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        {
+            return new ApplicationException("Unexpected error when creating reservation");
+        }
+
+        if (error.Params.Contains("StartDate", StringComparison.CurrentCultureIgnoreCase))
+        {
+            var currentDate = DateTime.Today;
+            var validFromDate = currentDate.AddMonths(-1).ToString("MM yyyy");
+            var validToDate = currentDate.AddMonths(2).ToString("MM yyyy");
+            var errorMessage = $"Training start date must be between the funding reservation dates {validFromDate} to {validToDate}";
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/Approvals/SFA.DAS.Approvals/Services/AutoReservationsService.cs:107]: That would be nice, but the Params check comes back in this form

`{
    "message": "The following parameters have failed validation (Parameter 'AccountId, StartDate')",
    "params": "AccountId, StartDate"
}`

therefore its not possible to without changing the API behaviour to return the actual error message raised within the API.

[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/Approvals/SFA.DAS.Approvals/Services/AutoReservationsService.cs:107]: We do have the option of returning the errors in another format used when validation rules fails, this alternative structure is used when the the reservation limit is reached or paused. However we will still need to map this error structure to the error structure used by commitments

File: src/Approvals/SFA.DAS.Approvals.UnitTests/Application/Cohorts/Commands/CreateCohortCommandHandlerTests.cs [hunk 2]
```diff
@@ -8,9 +9,11 @@
 using System.Threading.Tasks;
 using System.Threading;
 using FluentAssertions;
+using Microsoft.Azure.Amqp;
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/Approvals/SFA.DAS.Approvals.UnitTests/Application/Cohorts/Commands/CreateCohortCommandHandlerTests.cs:12]: done

File: src/Approvals/SFA.DAS.Approvals.UnitTests/Application/Cohorts/Commands/CreateCohortCommandHandlerTests.cs [hunk 3]
```diff
@@ -72,5 +77,79 @@ public async Task Handle_Cohort_Created()
...
+
+            try
+            {
+                await _handler.Handle(_request, CancellationToken.None);
+                Assert.Fail("Should not reach this point");
+            }
+            catch
+            {
+
+            }
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/Approvals/SFA.DAS.Approvals.UnitTests/Application/Cohorts/Commands/CreateCohortCommandHandlerTests.cs:151]: thx

**Review by ogrwill — APPROVED** [review]

**Pull request merged**